### ISSUE TITLE
Add more explanations of HitTestBehavior

### ISSUE THREAD

**Issue by fralways** [flutter/flutter#18450]
Internal: b/155674981

When I use `GestureDetector` and put `behavior` = `translucent`, I cannot tap on anything bellow it. Also when I use `GestureDetector` and put behavior lets say `opaque` and have `RawMaterialButton` as a `child`, taps on button are registered on `Button` but not on `GestureDetector` and expected behavior in this case would be that taps do not pass to the child.
Example:
```dart
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      home: new MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: new Stack(
        children: <Widget>[
          new Positioned(
            left: 0.0,
            right: 0.0,
            top: 200.0,
            bottom: 200.0,
            child: new TextField(
              textAlign: TextAlign.center,
              style: const TextStyle(
                color: Colors.blue,
              ),
            ),
          ),
          new Positioned.fill(
            child: new GestureDetector(
              behavior: HitTestBehavior.translucent,
              onTap: (){
                print("Clicked gesture");
              },
            ),
          ),

//          new Positioned.fill(
//            child: new GestureDetector(
//              behavior: HitTestBehavior.opaque,
//              onTap: (){
//                print("Clicked gesture");
//              },
//              child: new RawMaterialButton(
//                child: new Text(
//                  'BUTTON',
//                  textAlign: TextAlign.center,
//                ),
//                onPressed: (){
//                  print("Clicked button");
//                },
//              ),
//            ),
//          ),
        ],
      ),
    );
  }
}
```
In this example, I would expect that GestureDetector receives tap input, but also that TextField gets focus, but what happens is that taps are registered in GestureDetector, but nothing passes to the TextField. If I comment out GestureDetector part, TextField works.
Further, you can comment out first two widgets in stack and uncomment last one to check other issue- when GestureDetector has child. I would expect that no tap is passed to child but child receives input and GestureDetector does not.

```
flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, v0.5.5-pre.10, on Mac OS X 10.13.4 17E202, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.4)
[✓] Android Studio (version 3.1)
[!] IntelliJ IDEA Ultimate Edition (version 2018.1.2)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] Connected devices (2 available)

```

**Comment by zoechi** [comment]
@Hixie

**Comment by Hixie** [comment]
opaque just means the entire size of the gesture detector is the hit region, it doesn't stop the child from being hit. To block hits on children use an IgnorePointer or AbsorbPointer.

translucent means that the hit testing will traverse below. However, if both widgets are listening for the same gesture, in this case Tap, only one will "win". You can tell that both are getting the pointers though by listening to onTapDown instead of onTap.

**Comment by fralways** [comment]
Oh my, I completely misunderstand this. I guess adding those comments to the HitTestBehavior enum description would prevent other people from misunderstanding as well. I'm wondering now how can we register same event on multiple widgets?

**Comment by Hixie** [comment]
Use a PointerListener instead of a GestureDetector.

**Comment by zoechi** [comment]
@Hixie do you mean https://docs.flutter.io/flutter/rendering/RenderPointerListener-class.html?
Flutter.io doesn't find anything with `PointerListener`

**Comment by Hixie** [comment]
Sorry it's just called `Listener` at the widget level. (It's the one that creates a RenderPointerListener.)

**Comment by kaustavjaiswal** [comment]
@Hixie I am facing the same issue, I have a Stack widget on the main app screen and the last widget on the stack intercepts all the onTap calls, any example on how to make the onTap reach the lower widgets, positioned below in the stack would be extremely helpful.

**Comment by grigy** [comment]
I have a similar issue. I want to wrap the child of a [Dismissible](https://docs.flutter.io/flutter/widgets/Dismissible-class.html) widget with a GuestureDetector so that I can catch the horizontal swipe events and add additional functionality. But when I add the GestureDetector the Dismissible stops to respond to swipe action.

**Comment by wwwdata** [comment]
@kaustavjaiswal did you find a solution for this problem? Also have this issue.. The only solution for me is to put things that need gestures on top, so at the end of the stack list.

**Comment by kaustavjaiswal** [comment]
@wwwdata I am doing the same thing today.

https://flutter.io/gestures/

The gesture disambiguation section here helps to understand how flutter is treating these gestures.

**Comment by Tzxhy** [comment]
> @Hixie I am facing the same issue, I have a Stack widget on the main app screen and the last widget on the stack intercepts all the onTap calls, any example on how to make the onTap reach the lower widgets, positioned lower in the stack would be extremely helpful.

For now, 2020, I just saw this issue yesterday and I had a closer look at the source code about `Stack`, and finally located the problem. You can use the code below:

```dart



import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class StackWithAllChildrenReceiveEvents extends Stack {

  StackWithAllChildrenReceiveEvents({
    Key key,
    AlignmentDirectional alignment = AlignmentDirectional.topStart,
    TextDirection textDirection = TextDirection.ltr,
    StackFit fit = StackFit.loose,
    Overflow overflow = Overflow.clip,
    List<Widget> children = const <Widget>[],
  }) : super(
    key: key,
    alignment: alignment,
    textDirection: textDirection,
    fit: fit,
    overflow: overflow,
    children: children,
  );


  @override
  RenderStack createRenderObject(BuildContext context) {
    return RenderStackWithAllChildrenReceiveEvents(
      alignment: alignment,
      textDirection: textDirection ?? Directionality.of(context),
      fit: fit,
      overflow: overflow,
    );
  }

  @override
  void updateRenderObject(BuildContext context, RenderStackAllC renderObject) {
    renderObject
      ..alignment = alignment
      ..textDirection = textDirection ?? Directionality.of(context)
      ..fit = fit
      ..overflow = overflow;
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<AlignmentGeometry>('alignment', alignment));
    properties.add(EnumProperty<TextDirection>('textDirection', textDirection, defaultValue: null));
    properties.add(EnumProperty<StackFit>('fit', fit));
    properties.add(EnumProperty<Overflow>('overflow', overflow));
  }

}

class RenderStackWithAllChildrenReceiveEvents extends RenderStack {
  RenderStackWithAllChildrenReceiveEvents({
    List<RenderBox> children,
    AlignmentGeometry alignment = AlignmentDirectional.topStart,
    TextDirection textDirection,
    StackFit fit = StackFit.loose,
    Overflow overflow = Overflow.clip,
  }): super(
      alignment: alignment,
      textDirection: textDirection,
      fit: fit,
      overflow: overflow,
  );

  bool allCdefaultHitTestChildren(HitTestResult result, { Offset position }) {
    // the x, y parameters have the top left of the node's box as the origin
    RenderBox child = lastChild;
    while (child != null) {
      final StackParentData childParentData = child.parentData;
      child.hitTest(result, position: position - childParentData.offset);
      child = childParentData.previousSibling;
    }
    return false;
  }

  @override
  bool hitTestChildren(HitTestResult result, { Offset position }) {
    return allCdefaultHitTestChildren(result, position: position);
  }
}

```
The key code is `allCdefaultHitTestChildren`, you can diff it with `defaultHitTestChildren`. Use it just in which case you need all the children receive events not just last.

**Comment by stuz5000** [comment]
I'm totally confused by what to expect here.

With:
``` 
 - GestureDetector (onPanDown)
   - Container (green)
     -  GestureDetector (onPanDown, HitTestBehavior.opaque)
        - Container (red)
```
... both get the event. If I switch to:
``` 
 - Listener (onPointerDown)
   - Container (green)
     -  GestureDetector (onPanDown, HitTestBehavior.opaque)
        - Container (red)
```
... again, both receive the events.

Suppose I wanted to:
 
 - eagerly detect double taps in the green box AND drag events in the green box
 - unless a gesture started in an opaque child, such as a button. 

(NB. I can't used two GestureDetector detector how here because it waits until the completion of a feature before issuing the down event, which feels like UI jank to the user.)

Both of these both receive down events. Is it expected?
```
Listener(
    onPointerDown: (_) {
      print("Green");
    },
      child: Container(
        color: Colors.green,
        width: 300.0,
        height: 300.0,
        child: Center(
          child: GestureDetector(
            behavior: HitTestBehavior.opaque,
            onPanDown: (_) {
              print("Red");
            },
            child: Container(
              color: Colors.red,
              width: 50.0,
              height: 50.0,
            ),
          ),
        ),
      ),
  );
```

```
GestureDetector(
    onPanDown: (_) {
      print("Green");
    },
      child: Container(
        color: Colors.green,
        width: 300.0,
        height: 300.0,
        child: Center(
          child: GestureDetector(
            behavior: HitTestBehavior.opaque,
            onPanDown: (_) {
              print("Red");
            },
            child: Container(
              color: Colors.red,
              width: 50.0,
              height: 50.0,
            ),
          ),
        ),
      ),
  );
```

**Comment by elorusso** [comment]
I'm also having issues with this.  I have a stack of widgets.  One of the top widgets is translucent and placed above a Webview.  I don't want the Webview receiving touches underneath the translucent widget, so it seems that HitTestBehavior.opaque should do exactly this, based on the description: 

> Opaque targets can be hit by hit tests, causing them to both receive
> events within their bounds and prevent targets visually behind them from
> also receiving events.

This simply doesn't work. Very frustrating.

**Comment by mehmetf** [comment]
This is being asked internally as well and the behavior is confusing. Even if I use different handlers, the top one in the stack always seems to win:

```dart
      child: Stack(
          children: <Widget>[
            GestureDetector(
              onTapUp: (_) => print('Bottom Detector!'),
              behavior: HitTestBehavior.translucent,
              child: Container(width: 300, height: 200, color: Colors.red),
            ),
            GestureDetector(
              onTapDown: (_) => print('Top Detector!'),
              behavior: HitTestBehavior.translucent,
              child: Container(width: 200, height: 300, color: Colors.green),
            ),
          ],
        ),
```

@shihaohong is there a way to get this to work? I am assuming hit test behavior is not working here because these are siblings even though they overlap. I assume some parent widget needs to set the hit test behavior instead.

**Comment by stuz5000** [comment]
To be clear - there’s are also issues here with pan events not respecting opaque. (See my example above)

**Comment by shihaohong** [comment]
I've been poking around with `HitTestBehavior` all morning and I'm pretty confused myself. The only thing I could figure out was that when the `Container` has no color (completely transparent) and when the bottom widget is a `Listener` (When I use a `GestureDetector`, it never goes through). 

These are my observations:

**Tap on overlapping area results in both prints:**
```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Stack(
        children: <Widget>[
          Listener(
            onPointerUp: (_) => print('Bottom Detector!'),
            child: Container(width: 300, height: 200, color: Colors.red),
          ),
          Container(
            width: 200, 
            height: 300, 
            child: GestureDetector(
              onTapDown: (_) => print('Top Detector!'),
              behavior: HitTestBehavior.translucent, // translucent
            ),
          ),
        ],
      ),
    );
  }
}
```

**Tap on overlapping area results in only top print when top widget is non-transparent:**
```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Stack(
        children: <Widget>[
          Listener(
            onPointerUp: (_) => print('Bottom Detector!'),
            child: Container(width: 300, height: 200, color: Colors.red),
          ),
          Container(
            width: 200, 
            height: 300, 
            color: Colors.orange, // adding this line results in Listener not receiving pointer events
            child: GestureDetector(
              onTapDown: (_) => print('Top Detector!'),
              behavior: HitTestBehavior.translucent, // translucent
            ),
          ),
        ],
      ),
    );
  }
}
```

**Tap on overlapping area results in only top print when HitTestBehavior.opaque and top widget is transparent:**
```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Stack(
        children: <Widget>[
          Listener(
            onPointerUp: (_) => print('Bottom Detector!'),
            child: Container(width: 300, height: 200, color: Colors.red),
          ),
          Container(
            width: 200, 
            height: 300, 
            // color: Colors.orange // no color, so transparent
            child: GestureDetector(
              onTapDown: (_) => print('Top Detector!'),
              behavior: HitTestBehavior.opaque, // opaque
            ),
          ),
        ],
      ),
    );
  }
}
```

**Comment by shihaohong** [comment]
Adding another another sample to help further with future investigations... I expect that since all Listeners have its behaviors set to `HitTestBehavior.opaque`, only one print statement would go through, but all three do.

```dart
import 'package:flutter/material.dart';

final Color darkBlue = Color.fromARGB(255, 18, 32, 47);

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyWidget(),
    );
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Container(
          color: Colors.blue,
          child: Listener(
            behavior: HitTestBehavior.opaque,            
            onPointerDown: (_) => print('bottom'),
            child: Listener(
              behavior: HitTestBehavior.opaque,
              onPointerDown: (_) => print('middle'),
              child: Listener(
                behavior: HitTestBehavior.opaque,
                onPointerDown: (_) => print('top')
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

**Comment by Hixie** [comment]
These are children of each other so they all get hit.

**Comment by shihaohong** [comment]
Does `HitTestBehavior` only matter in the context of `Stack`s and sibling widgets?

**Comment by goderbauer** [comment]
@mehmetf Your example is indeed a little confusing.

The Container introduces a ColoredBox, whose RenderObject has opaque HitTestingBehavior hard-coded, see here:

https://github.com/flutter/flutter/blob/bbc0cc07f685b64fb7a191810b7299db44b46ec6/packages/flutter/lib/src/widgets/basic.dart#L7107-L7110

The surrounding GestureDetector may have translucent HitTestBehavior, but that's overwritten by the child's opaque behavior:

https://github.com/flutter/flutter/blob/bbc0cc07f685b64fb7a191810b7299db44b46ec6/packages/flutter/lib/src/rendering/proxy_box.dart#L167

`hitTarget` evaluates to true there (because of the opaque child) preventing other children in the stack from getting the hit:

https://github.com/flutter/flutter/blob/bbc0cc07f685b64fb7a191810b7299db44b46ec6/packages/flutter/lib/src/rendering/box.dart#L2485-L2503

`isHit` is true for the first child, causing the Stack to skip hit-testing all other children. The ColoredBox in your example prevents the second GestureDetector from getting a chance in participating in hit testing at all.

Now, if you remove the color from the second container, both can participate in hit-testing, but your bottom detector will still not see its `onTapUp` fire. That's because `onTapUp` only fires when a gesturedetector wins the gesture arena. In your case, there's a tie between the two gesture detectors, which is resolved by making the first one to enter win (the top one in this case). For the top gesture detector you will see `onTapDown` and `onTapUp` fire. For the bottom one you will see `onTapDown` and `onTapCancel` fire.

If you remove the color of the second container AND switch both to `onTapDown` you should see them both respond.

(This is just to explain what's currently happening, not sure if it makes sense that the color of a container interferes with hit testing in this way.)

**Comment by goderbauer** [comment]
One thing I am curious about: What are people's use cases? It seems odd to me to have two overlapping gesture detectors of the same kind and expect them to both react. What's that used for?

**Comment by stuz5000** [comment]
For me, I want to be able to tap (onDown) a custom widget to change its
state. Tapping the background
should return it to the original state.

**Comment by goderbauer** [comment]
@stuz5000 That should just work since in that example you never want both gesture detectors to fire simultaneous. Can you post some runnable code that showcases the problem you're facing with that?

**Comment by stuz5000** [comment]
See my example above.

**Comment by goderbauer** [comment]
Oh, you have them organized as children of each other, not as siblings in a stack. In your case, what Hixie said above (https://github.com/flutter/flutter/issues/18450#issuecomment-601840123) applies. If you were to organize them in a stack you could get the behavior you want.

**Comment by stuz5000** [comment]
To be clear - what I expected:

 - siblings in a stack will get both the events if the top widget is transparent (and is working as expected)
 - siblings in a stack will get one event when the top-most is opaque (and is working as expected)
 - parent-child to have the only the child have the event if the child is tapped, or the parent if the parent (and not the child is tapped)

But in fact:
 - widgets in parent-child relationship get both events whatever the opacity

So it is resolvable in edge cases with stacks. This doesn't make life easier except in edge cases since we can't have all our widgets in a Stack for every problem and the go-to organizing pattern throughout Flutter is almost all parent-child.

If we put all our stuff in stack, we're now self-managing layout everywhere. Is this parent-child behavior by design, or an oversight?

```
import 'package:flutter/material.dart';
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Welcome to Flutter',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Welcome to Flutter'),
        ),
        body: Center( child: Column(children: [
          // Parent-child -- two events when red is tapped
          Listener(
            onPointerDown: (_) {  print("Green");  },
            child: Container(  color: Colors.green,  width: 100.0, height: 100.0,  child: Center(
              child: GestureDetector(
                behavior: HitTestBehavior.opaque,
                onPanDown: (_) {  print("Red");   },
                child: Container(color: Colors.red, width: 50.0,height: 50.0,  ),
              ),
            ),
            ),
          ),
          SizedBox(width:10,height: 10),
          // Stacked -- one event when red is tapped
          Stack(
            children: [
              Listener(
                  onPointerDown: (_) { print("Green");},
                  child: Container(color: Colors.green, width: 100.0, height: 100.0)),
              Positioned(
                left:25, top:25,
                child: GestureDetector(
                    behavior: HitTestBehavior.opaque,
                    onPanDown: (_) {  print("Red");   },
                    child: Container(color: Colors.red, width: 50.0,height: 50.0,  )
              ))
            ],
          ),
        ])
        ),
      ),
    );
  }
}
```

**Comment by om-ha** [comment]
@stuz5000 your answer really helped, I ended up having something like this:

``` dart
return Stack(
  children: <Widget>[
    SomeOtherNestedWidgetWithGestureDetectors(),
    Listener(
      behavior: HitTestBehavior.translucent,
      onPointerDown: (pointerDownEvent) {
        // PUT YOUR CODE HERE
      },
    ),
  ],
);
```

My use case consists of putting a full screen gesture detector that bleeds/forwards/passes touches to the views behind it.
Using `GestureDetector` with `Translucent` `HitTestBehavior` did not work. I had to use Listener and, like GestureDetector`, set its `behavior` to `Translucent`.

**Comment by stuz5000** [comment]
Bump. It sure could use some documentation. Can anyone from the Flutter team make it clear what the intended behavior is:

```
+ Listener/GestureDetector
   + Container  # A
      + Listener/GestureDetector
          + Container  # B
```
(see https://github.com/flutter/flutter/issues/18450#issuecomment-583212161)

Both A and B receive on-down events if B is visually-on-top of A. Bug or by design?

**Comment by michaelniedermayr** [comment]
@Tzxhy Your modified stack class `StackWithAllChildrenReceiveEvents` works like a charm. Thx a lot.

**Comment by xJon** [comment]
Indeed @Tzxhy's modified Stack class fixes the issue.

**Comment by stuz5000** [comment]
@xJon  -- it is only a fix if you use stacks for your layout.

**Comment by xJon** [comment]
@stuz5000 You're right; I happened to use Stacks for my layout.

**Comment by DenisBarzanov** [comment]
> @mehmetf Your example is indeed a little confusing.
> 
> The Container introduces a ColoredBox, whose RenderObject has opaque HitTestingBehavior hard-coded, see here:
> 
> https://github.com/flutter/flutter/blob/bbc0cc07f685b64fb7a191810b7299db44b46ec6/packages/flutter/lib/src/widgets/basic.dart#L7107-L7110
> 
> The surrounding GestureDetector may have translucent HitTestBehavior, but that's overwritten by the child's opaque behavior:
> 
> https://github.com/flutter/flutter/blob/bbc0cc07f685b64fb7a191810b7299db44b46ec6/packages/flutter/lib/src/rendering/proxy_box.dart#L167
> 
> `hitTarget` evaluates to true there (because of the opaque child) preventing other children in the stack from getting the hit:
> 
> https://github.com/flutter/flutter/blob/bbc0cc07f685b64fb7a191810b7299db44b46ec6/packages/flutter/lib/src/rendering/box.dart#L2485-L2503
> 
> `isHit` is true for the first child, causing the Stack to skip hit-testing all other children. The ColoredBox in your example prevents the second GestureDetector from getting a chance in participating in hit testing at all.
> 
> Now, if you remove the color from the second container, both can participate in hit-testing, but your bottom detector will still not see its `onTapUp` fire. That's because `onTapUp` only fires when a gesturedetector wins the gesture arena. In your case, there's a tie between the two gesture detectors, which is resolved by making the first one to enter win (the top one in this case). For the top gesture detector you will see `onTapDown` and `onTapUp` fire. For the bottom one you will see `onTapDown` and `onTapCancel` fire.
> 
> If you remove the color of the second container AND switch both to `onTapDown` you should see them both respond.
> 
> (This is just to explain what's currently happening, not sure if it makes sense that the color of a container interferes with hit testing in this way.)

Oh my god someone really knows the Flutter codebase here wow

**Comment by alexeyselina** [comment]
Does somebody know how to fix StackWithAllChildrenReceiveEvents class? 
I'm totally confused with overriding of 
'bool hitTestChildren(BoxHitTestResult result, { required Offset position }) '

When I'm trying to overload it offers me this method:
`@override
  bool hitTestChildren(BoxHitTestResult result, {Offset}) {
    // TODO: implement hitTestChildren
    return super.hitTestChildren(result, Offset);
  }`

I have 2.10.3 Dart version but the syntax causes compilation issues.

**Comment by xJon** [comment]
@alexeyselina This is how I have it:
```dart
@override
  bool hitTestChildren(HitTestResult result, {Offset position}) {
    return allCdefaultHitTestChildren(result, position: position);
  }
```

**Comment by h65wang** [comment]
FWIW, this example correctly showcases all 3 HitTestBehaviors:

```
Stack(
  children: <Widget>[
    GestureDetector(
      onTapDown: (_) => print('Bottom Container'),
      child: Container(width: 200, height: 200, color: Colors.blue),
    ),
    GestureDetector(
      /* change the behavior property results in 3 different behaviors:
         (try long pressing on the blue square when running the app)

       * translucent: gestures pass through, so both events will trigger.

       * opaque: top (this container) blocks the bottom container,
         so only long press event on this container is triggered.

       * deferToChild: this child (transparent container) won't accept clicks,
         so only the bottom container will be triggered. */
      behavior: HitTestBehavior.translucent,
      onLongPressStart: (_) => print('Long press of top container!'),
      child: Container(width: 200, height: 200),
    ),
  ],
)
```

**Comment by stuz5000** [comment]
> FWIW, this example correctly showcases all 3 HitTestBehaviors, remember to long press when testing:

Yes... works for Stacks, but not for nested widgets, which I would think is the very common cases.

**Comment by stuz5000** [comment]
A simple workaround to the bug of duplicate event when mixing GeatureRecognizer and Listener in parent child relationships is to wrap the Listener with an additional

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Welcome to Flutter',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Welcome to Flutter'),
        ),
        body: Center(
            child: Column(children: [
          // Parent-child -- two events when red is tapped
          GestureDetector(
            onTap: () {
              print("Green");
            },
            child: Container(
              color: Colors.green,
              width: 100.0,
              height: 100.0,
              child: Center(
                child: GestureDetector(    //  <<<<< Add this junk
                  onTap: () {},                    //  <<<<< Add this junk
                  child: Listener(
                    onPointerDown: (_) {
                      print("Red");
                    },
                    child: Container(
                      color: Colors.red,
                      width: 50.0,
                      height: 50.0,
                    ),
                  ),
                ),
              ),
            ),
          ),
        ])),
      ),
    );
  }
}
```

Thanks to: https://stackoverflow.com/questions/65572920/how-to-avoid-duplicate-events-when-combining-gesturedetector-and-listener-in-flu

**Comment by giorgio79** [comment]
> > @Hixie I am facing the same issue, I have a Stack widget on the main app screen and the last widget on the stack intercepts all the onTap calls, any example on how to make the onTap reach the lower widgets, positioned lower in the stack would be extremely helpful.
> 
> @kaustavjaiswal. For now, 2020, I just saw this issue yesterday and I had a closer look at the source code about `Stack`, and finally located the problem. You can use the code below:
> 
> ```dart
> import 'package:flutter/material.dart';
> import 'package:flutter/rendering.dart';
> 
> class StackWithAllChildrenReceiveEvents extends Stack {
> 
>   StackWithAllChildrenReceiveEvents({
>     Key key,
>     AlignmentDirectional alignment = AlignmentDirectional.topStart,
>     TextDirection textDirection = TextDirection.ltr,
>     StackFit fit = StackFit.loose,
>     Overflow overflow = Overflow.clip,
>     List<Widget> children = const <Widget>[],
>   }) : super(
>     key: key,
>     alignment: alignment,
>     textDirection: textDirection,
>     fit: fit,
>     overflow: overflow,
>     children: children,
>   );
> 
> 
>   @override
>   RenderStackWithAllChildrenReceiveEvents createRenderObject(BuildContext context) {
>     return RenderStackWithAllChildrenReceiveEvents(
>       alignment: alignment,
>       textDirection: textDirection ?? Directionality.of(context),
>       fit: fit,
>       overflow: overflow,
>     );
>   }
> 
>   @override
>   void updateRenderObject(BuildContext context, RenderStackWithAllChildrenReceiveEvents renderObject) {
>     renderObject
>       ..alignment = alignment
>       ..textDirection = textDirection ?? Directionality.of(context)
>       ..fit = fit
>       ..overflow = overflow;
>   }
> 
>   @override
>   void debugFillProperties(DiagnosticPropertiesBuilder properties) {
>     super.debugFillProperties(properties);
>     properties.add(DiagnosticsProperty<AlignmentGeometry>('alignment', alignment));
>     properties.add(EnumProperty<TextDirection>('textDirection', textDirection, defaultValue: null));
>     properties.add(EnumProperty<StackFit>('fit', fit));
>     properties.add(EnumProperty<Overflow>('overflow', overflow));
>   }
> 
> }
> 
> class RenderStackWithAllChildrenReceiveEvents extends RenderStack {
>   RenderStackWithAllChildrenReceiveEvents({
>     List<RenderBox> children,
>     AlignmentGeometry alignment = AlignmentDirectional.topStart,
>     TextDirection textDirection,
>     StackFit fit = StackFit.loose,
>     Overflow overflow = Overflow.clip,
>   }): super(
>       alignment: alignment,
>       textDirection: textDirection,
>       fit: fit,
>       overflow: overflow,
>   );
> 
>   bool allCdefaultHitTestChildren(HitTestResult result, { Offset position }) {
>     // the x, y parameters have the top left of the node's box as the origin
>     RenderBox child = lastChild;
>     while (child != null) {
>       final StackParentData childParentData = child.parentData;
>       child.hitTest(result, position: position - childParentData.offset);
>       child = childParentData.previousSibling;
>     }
>     return false;
>   }
> 
>   @override
>   bool hitTestChildren(HitTestResult result, { Offset position }) {
>     return allCdefaultHitTestChildren(result, position: position);
>   }
> }
> ```
> 
> The key code is `allCdefaultHitTestChildren`, you can diff it with `defaultHitTestChildren`. Use it just in which case you need all the children receive events not just last.

Testing StackWithAllChildrenReceiveEvents 
I notice overflow is getting an error, saying the setter is not defined using flutter stable channel

```
flutter doctor -v
[√] Flutter (Channel stable, 1.22.6, on Microsoft Windows [Version 10.0.16299.2166], locale hu-HU)
    • Flutter version 1.22.6 at C:\flutter
    • Framework revision 9b2d32b605 (10 days ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5


[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\gchityil\AppData\Local\Android\Sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: C:\Program Files\Android\Android Studio1\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.


```

**Comment by giorgio79** [comment]
I completed testing StackWithAllChildrenReceiveEvents for the issue with sample at https://github.com/flutter/flutter/issues/75086#issuecomment-770422447

The bottom child in the Stack is now receiving the events, so all looks good! Thanks a lot!! :) 
I had to remove the EagerGestureRecognizer from the bottom child, GoogleMap, since with that none of the children above the GoogleMap were getting gestures. Without EagerGestureRecognizer it works like magic.
